### ISSUE TITLE
Makes processing robust against imported NSRL databases with product code inconsistencies

### ISSUE THREAD

**Issue by leosol** [sepinf-inc/IPED#611]
While trying to import the latest NSRL database, IPED 3.18.6 returned a NullPointer Exception error.

Debugging the KFFTask.java class I've noticed that there is no check for Null values, which caused the error.

Check the image! Just tell me if there is anything I can do to help you guys:

![Bug-iped-3 18 6-KFFTask](https://user-images.githubusercontent.com/4167122/122601990-7d19d600-d048-11eb-93a3-7e3c226ec671.jpg)

**Comment by leosol** [comment]
attr was under zero and prodcuts.get(Math.abs(attr) returned null.
Unfortunately I'm not so sure if the solution would be to change or (||) for and (&&) or just assert that product should not be null...

**Comment by tc-wleite** [comment]
Hi @leosol! 
Are you getting this error while importing or while processing after importing?
Which file are you trying to import?
I will try to run here and see if I can figure it out...

About the source code, as far as I know, NSRL product codes are always filled with valid values, so it should not be null at that point of the code. 
Probably there is something else going on...

**Comment by lfcnassif** [comment]
@leosol please post the full error stacktrace.

**Comment by lfcnassif** [comment]
This was already reported before https://github.com/sepinf-inc/IPED/discussions/368#discussioncomment-380462

But the user used a non stardard NSRL database that time, possibly non compliant. If this happens with default NSRL, we should fix.

**Comment by lfcnassif** [comment]
@leosol a snapshot with a possible fix will be built in 5min in https://github.com/sepinf-inc/IPED/actions/runs/950889514

Please check if it solves your issue.

**Comment by leosol** [comment]
On monday I will send the full stacktrace, sorry for forgetting it. We did some tests with a custom database and with the NSRL one. , I guess it happened with the latter, but I'm not sure. On monday we check everything and test the snapshot build. Thank you guys for being so fast!

**Comment by lfcnassif** [comment]
So please check which hash database contains negative product code values, thanks.

**Comment by leosol** [comment]
Sure

**Comment by DHoelz** [comment]
Just to complete @leosol answer as he is helping me out with this problem.

The problem happens while **processing** with the latest NSRL database with **no** modifications (RDS Version 2.73 - June 2021). I also verified the sha-1 hash of the NSRL files with the _signatures.txt_ before importing. 

What I've tried so far:
1. Tried to process 2 different disk images (Encase format) and the problem happened with both of them.
2. Processed with an old **custom** database I use daily (twice the size of the normal NRSL database) and it worked normally with both images. This "old" database probably uses the NSRL database released earlier this year but i'm not 100% sure.

**Comment by tc-wleite** [comment]
> The problem happens while **processing** with the latest NSRL database with **no** modifications (RDS Version 2.73 - June 2021). I also verified the sha-1 hash of the NSRL files with the _signatures.txt_ before importing.

@DHoelz, thanks for the information!
I am downloading this new versions of NSRL and will test here to see if I can reproduced it.

As we changed how IPED deals with hashes databases for the next major release, I also would like to check if these new files are handled properly.

Which set did you imported: full, compact or minimal?
Were iOS and Android RDS also imported?

**Comment by DHoelz** [comment]
@tc-wleite I'm using the **Modern RDS  (minimal)**. I tried using only the Android RDS and got the same problem. Don't remember trying the iOS database by itself.

BTW, I just tried using the _snapshot_ and got the same error.

Thanks for the help!

**Comment by tc-wleite** [comment]
> @leosol a snapshot with a possible fix will be built in 5min in https://github.com/sepinf-inc/IPED/actions/runs/950889514
> 
> Please check if it solves your issue.

@lfcnassif, I **think** your fix should solve the reported error, but a few lines after the place you changed, there is another line that should cause a similar error:
 
```
String[] product = products.get(Math.abs(attr));
if (attr > 0 || (product != null && alertProducts.contains(product[0])))
{
    evidence.setExtraAttribute(KFF_STATUS, "alert"); //$NON-NLS-1$
} else {
    if (excludeKffIgnorable) {
        evidence.setToIgnore(true);
        synchronized (lock) {
            excluded++;
        }
    } else {
        evidence.setExtraAttribute(KFF_STATUS, "ignore"); //$NON-NLS-1$
    }
}
if (product != null) // <<<<------ I think this check is also necessary
    evidence.setExtraAttribute(KFF_GROUP, product[0] + " " + product[1]); //$NON-NLS-1$
```

**Comment by lfcnassif** [comment]
Thanks @DHoelz. Just confirmed latest Android NSRLFile.txt references several times at least 1 product code (456485) not present in NSRLProd.txt file. So, this is an inconsistency in latest NSRL 2.73.

> I **think** your fix should solve the reported error, but a few lines after the place you changed, there is another line that may cause a similar error:

uh... that's the result of not trying to reproduce the bug and check the fix...

I will make 3.18.x code robust against NSRL inconsistencies. @tc-wleite if you could check if master branch is robust against this NSRL issue, that would be great.

**Comment by tc-wleite** [comment]
NSRL files are broken.

The first lines of **NSRLFile.txt** (RDS Version 2.73 - June 2021 - full set):
```
"SHA-1","MD5","CRC32","FileName","FileSize","ProductCode","OpSystemCode","SpecialCode"
"0000001FFEF4BE312BAB534ECA7AEAA3E4684D85","344428FA4BA313712E4CA9B16D089AC4","7516A25F",".text._ZNSt14overflow_errorC1ERKSs",33,219181,"362",""
"00000052A9EEEC6C8348CFB2AEA77BC1FBF8D239","F46CA74CA3D89E9D3CF8D8E5CD77842D","2F9CC135","__DATA__mod_init_func",772,448056,"362",""
```

This second line has **ProductCode = 448056**.
But there is no such ProductCode in *NSRLProd.txt*.
It happens with a lot of records (at least 3 million).

**Comment by lfcnassif** [comment]
hohoho I was faster :-)

**Comment by tc-wleite** [comment]
> I will make 3.18.x code robust against NSRL inconsistencies. @tc-wleite if you could check if master branch is robust against this NSRL issue, that would be great.

Just tested and it isn't.
I reported the issue to NIST.
And will make the changes to the master to handle this type of inconsistency tomorrow.
Anyway, I guess (hope) they will fix it, as it is a serious flaw, and never happen (even with a single record) in previous versions.

**Comment by tc-wleite** [comment]
Just to be more clear, with the master there is an error while importing these new NSRL files.
I will first fix this issue, then check if processing is affected.

**Comment by tc-wleite** [comment]
Just updating the actual numbers, this issue affects ~6.32 million records out of 166 million (in the full set), which is a lot.

**Comment by tc-wleite** [comment]
Found another problem in the files, when I tried to import Android set with the master branch...
It has one malformed line (unescaped quotes in the middle of field) in **NSRLProd.txt**:
`216417,"Nexus 6P ("angler") binaries for Android (NMF26F)","unspecified","1160","1618","English","Drivers"`

This causes an error in the CSV parsing library we are using.
I don't think we should handle this kind of issue.
I edited this line and it was enough to make the file valid.

The main issue (product codes in NSRLFile.txt that do not exist in NSRLProd.txt) affects ~4.81 million out of 29.6 million records in Android data set.

**Comment by lfcnassif** [comment]
This NSRL is really broken, I also think they Will fix after your report @tc-wleite, thank you

**Comment by lfcnassif** [comment]
@leosol @DHoelz please test this snapshot https://github.com/sepinf-inc/IPED/actions/runs/957295451

**Comment by lfcnassif** [comment]
I also added warnings when importing NSRL format with product code inconsistencies.

**Comment by DHoelz** [comment]
@lfcnassif, I made a quick test and it seems to be working fine. 

Thanks!

**Comment by lfcnassif** [comment]
Thanks @DHoelz, so I am closing this.

**Comment by tc-wleite** [comment]
> This NSRL is really broken, I also think they Will fix after your report @tc-wleite, thank you

Just a quick update, got a reply from NIST saying that they will release a new version (2.73.1) fixing this issue as soon as possible:

_I've also got about 6.3 million rows (6,319,360 to be exact) that have incorrect product codes. This number happens to be the exact number of new lines we've added to NSRLFile since the previous RDS release and is in fact only an issue with all new rows added in the current release. The value in the product code column for these incorrect rows is another internal identifying number we use for software packages and was published in error._

**Comment by lfcnassif** [comment]
> is in fact only an issue with all new rows added in the current release

good to know

**Comment by tc-wleite** [comment]
NIST released a new version (2.73.1):
https://www.nist.gov/itl/ssd/software-quality-group/national-software-reference-library-nsrl/nsrl-download/current-rds
I am downloading and will try to import here (with the master version).

And not about the issue:
https://www.nist.gov/itl/ssd/software-quality-group/june-2021-nist-nsrl-rds-273-publication-error

**Comment by tc-wleite** [comment]
> I am downloading and will try to import here (with the master version).

As expected, the version 2.73.1 is correct, without any product code inconsistency or malformed CSV.
